File: kotlin/space-age/src/main/kotlin/SpaceAge.kt
import java.math.BigDecimal
import java.math.RoundingMode

class SpaceAge(val seconds: Long) {
    private val preciseEarthAge = seconds.div(31557600.0)

    fun onEarth() = format(preciseEarthAge)

    fun onMercury() = format(preciseEarthAge.div(0.2408467))

    fun onVenus() = format(preciseEarthAge.div(0.61519726))

    fun onMars() = format(preciseEarthAge.div(1.8808158))

    fun onJupiter() = format(preciseEarthAge.div(11.862615))

    fun onSaturn() = format(preciseEarthAge.div(29.447498))

    fun onUranus() = format(preciseEarthAge.div(84.016846))

    fun onNeptune() = format(preciseEarthAge.div(164.79132))

    private fun format(highPrecisionNumber: Double): Double {
        return BigDecimal(highPrecisionNumber).setScale(2, RoundingMode.HALF_UP).toDouble()
    }
}

File: scala/space-age/src/main/scala/SpaceAge.scala
import scala.math.BigDecimal.RoundingMode.HALF_UP

case class SpaceAge(seconds: Long) {
  private val PreciseEarthOrbitalPeriod = seconds / 31557600d

  val onEarth: Double = ageFor(1d)
  val onMercury: Double = ageFor(0.2408467d)
  val onVenus: Double = ageFor(0.61519726d)
  val onMars: Double = ageFor(1.8808158)
  val onJupiter: Double = ageFor(11.862615d)
  val onSaturn: Double = ageFor(29.447498d)
  val onUranus: Double = ageFor(84.016846d)
  val onNeptune: Double = ageFor(164.79132d)

  private def ageFor(orbitalPeriodInEarthYears: Double) = {
    format(PreciseEarthOrbitalPeriod / orbitalPeriodInEarthYears)
  }

  private def format(double: Double) = {
    BigDecimal(double).setScale(2, HALF_UP).toDouble
  }
}
